umh-core/pkg/service/filesystem/buffered_filesystem.go

```diff
@@ -0,0 +1,786 @@
+		}
+		if err := bs.base.WriteFile(ctx, path, chg.content, chg.perm); err != nil {
+			return fmt.Errorf("failed to write file: %w", err)
+		}
+		delete(bs.changed, path)
+	}
+
+	// Step 3 & 4: Handle removals in reverse order (deepest paths first)
+	var filesToRemove []string
+	var dirsToRemove []string
+	for path, chg := range bs.changed {
+		if chg != nil && chg.removed {
+			if chg.wasDir {
+				dirsToRemove = append(dirsToRemove, path)
+			} else {
+				filesToRemove = append(filesToRemove, path)
+			}
+		}
+	}
+
+	// Sort paths by length in descending order
+	sort.Slice(filesToRemove, func(i, j int) bool {
+		return len(filesToRemove[i]) > len(filesToRemove[j])
+	})
+	sort.Slice(dirsToRemove, func(i, j int) bool {
+		return len(dirsToRemove[i]) > len(dirsToRemove[j])
+	})
+
+	// First remove files
+	for _, path := range filesToRemove {
+		if err := bs.base.Remove(ctx, path); err != nil {
+			return fmt.Errorf("failed to remove file: %w", err)
+		}
+		delete(bs.files, path)
+		delete(bs.changed, path)
+	}
+
+	// Then remove directories
+	for _, path := range dirsToRemove {
+		if err := bs.base.RemoveAll(ctx, path); err != nil {
+			return fmt.Errorf("failed to remove directory: %w", err)
+		}
+		delete(bs.files, path)
+		delete(bs.changed, path)
+	}
+
+	return nil
+}
+
+// =========================
+// Implementation of Service
+// =========================
+
+// EnsureDirectory creates the directory in-memory and marks it for creation on disk.
+func (bs *BufferedService) EnsureDirectory(ctx context.Context, path string) error {
+	bs.mu.Lock()
+	defer bs.mu.Unlock()
+
+	// If already exists, check if it is a directory.
+	if state, ok := bs.files[path]; ok {
+		if !state.isDir {
+			return fmt.Errorf("%s exists and is not a directory", path)
+		}
+		return nil
+	}
+
+	// Create new directory entry.
+	bs.files[path] = &fileState{
+		isDir:    true,
+		modTime:  time.Now(),
+		fileMode: os.ModeDir | 0755,
+		size:     0,
+	}
+	// Mark as changed so that SyncToDisk can create it on disk.
+	bs.changed[path] = &fileChange{
+		removed: false,
+		perm:    os.ModeDir | 0755,
+	}
+	return nil
+}
+
+// ReadFile returns the content from the in-memory cache only.
+// If not present, returns os.ErrNotExist.
+func (bs *BufferedService) ReadFile(ctx context.Context, path string) ([]byte, error) {
+	bs.mu.Lock()
+	defer bs.mu.Unlock()
+
+	st, ok := bs.files[path]
+	if !ok || st.isDir {
+		return nil, os.ErrNotExist
+	}
+	// If it's a large file (content is nil but size is set), return not exist
+	if st.content == nil && st.size > bs.maxFileSize {
+		return nil, os.ErrNotExist
+	}
+	// Return empty slice instead of nil if content is nil
+	if st.content == nil {
+		return []byte{}, nil
+	}
+	return st.content, nil
+}
+
+// WriteFile does not immediately write to disk; it marks the file as changed in memory.
+func (bs *BufferedService) WriteFile(ctx context.Context, path string, data []byte, perm os.FileMode) error {
```

Comment:
This does not check if the given permissions will allow us to write.
Additionally, if the file already exists, the existing permissions might prevent a write, which is not cached here.
Also user and groups are ignored.

Reply:
Same goes for all other file ops